libs/parallelism/executors/include/hpx/executors/scheduler_executor.hpp

```diff
@@ -0,0 +1,319 @@
+                        std::forward<decltype(predecessor)>(predecessor)),
+                    std::move(t));
+            };
+        }
+#endif
+    }    // namespace detail
+
+    ///////////////////////////////////////////////////////////////////////////
+    // A scheduler_executor wraps any P2300 scheduler and implements the
+    // executor functionalities for those.
+    template <typename BaseScheduler>
+    struct scheduler_executor
+    {
+        template <typename Scheduler,
+            typename Enable = std::enable_if_t<
+                hpx::execution::experimental::is_scheduler_v<Scheduler>>>
+        constexpr explicit scheduler_executor(Scheduler&& sched)
+          : sched_(std::forward<Scheduler>(sched))
+        {
+        }
+
+        /// \cond NOINTERNAL
+        constexpr bool operator==(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ == rhs.sched_;
+        }
+
+        constexpr bool operator!=(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ != rhs.sched_;
+        }
+
+        constexpr auto const& context() const noexcept
+        {
+            return *this;
+        }
+        /// \endcond
+
+        /// \cond NOINTERNAL
+        // Associate the parallel_execution_tag executor tag type as a default
+        // with this executor.
+        using execution_category = parallel_execution_tag;
+
+        // Associate the static_chunk_size executor parameters type as a default
+        // with this executor.
+        using executor_parameters_type = static_chunk_size;
+
+        template <typename T, typename... Ts>
+        using future_type = hpx::future<T>;
+
+        // NonBlockingOneWayExecutor interface
+        template <typename F, typename... Ts>
+        void post(F&& f, Ts&&... ts)
+        {
+            detach(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // OneWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) sync_execute(F&& f, Ts&&... ts)
+        {
+            return sync_wait(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // TwoWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) async_execute(F&& f, Ts&&... ts)
+        {
+            return make_future(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        template <typename F, typename Future, typename... Ts>
+        decltype(auto) then_execute(F&& f, Future&& predecessor, Ts&&... ts)
+        {
+            auto&& sched =
+                on(keep_future(std::forward<Future>(predecessor)), sched_);
+
+            return make_future(transform(std::move(sched),
```

Comment:
```suggestion
            auto&& predecessor_on_sched =
                on(keep_future(std::forward<Future>(predecessor)), sched_);

            return make_future(transform(std::move(predecessor_on_sched),
```
or something in that direction? I don't particularly like that `sched_` and `sched` are completely different things with almost the same name.